File: python/84.Largest Rectangle in Histogram.py
from typing import List
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        heights.append(0)
        max_area = 0
        i = 0
        while(i < len(heights)):
            # print(stack, max_area)
            if len(stack) == 0 or heights[i] >= heights[stack[len(stack) - 1]]:
                stack.append(i)
                i += 1
            else:
                idx = stack.pop()
                if len(stack) == 0:
                    max_area = max(max_area, heights[idx] * i)
                else:
                    max_area = max(max_area, heights[idx] * (i - stack[len(stack)-1] - 1))
        return max_area


if __name__ == "__main__":
    print(Solution().largestRectangleArea([2,1,5,6,2,3]))

File: cpp/cmake-build-debug/84.Largest Rectangle in Histogram.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int>_stack;
        int max_area = 0;
        int i = 0;
        heights.push_back(0);
        while(i < heights.size()){
            if (_stack.empty() || heights[i] > heights[_stack.top()]){
                _stack.push(i);
                i++;
            }
            else{
                int idx = _stack.top();
                _stack.pop();
                if (_stack.empty())
                    max_area = max(max_area, heights[idx] * i);
                else
                    max_area = max(max_area, heights[idx] * (i - _stack.top() - 1));
            }
        }
        return max_area;
    }
};